### ISSUE TITLE
fuse_as_graphmodule change the partitioned exported program input order

### ISSUE THREAD

**Issue by chenweng-quic** [pytorch/pytorch#164204]
### 🐛 Describe the bug

We notice the ExportedProgram partitioner may change input order.
For example, for inputs in origin EP sent to Partitioner is [input_ids, attention_mask].
However, the EP sent to backend preprocess become [attention_mask, input_ids].

After analysis, the root cause comes from one function in the torch side.
The workflow for partition is below:  
1. Executorch partitioner gives each node a tag, meaning they can be grouped into a partition.
2. The ExecuTorch partitioner will call torch fuse_as_graphmodule to fuse nodes with same tag.
3. fuse_as_graphmodule will follow the visit order from nodes passed, to create the new placeholder in new graph.
4. Since the passed in nodes first use attention_mask, the input order of ExportedProgram in graph become [attention_mask, input_ids]
<img width="1226" height="289" alt="Image" src="https://github.com/user-attachments/assets/9b7afadc-7cb3-4faa-89a6-ff9167a10b84" />
https://github.com/pytorch/pytorch/blob/2507ae63f293354170695fd20a5c5ce5f64e323d/torch/fx/passes/utils/fuser_utils.py

I prepare a script to reproducible the issue: https://gist.github.com/chenweng-quic/80bf4b957552a87d792c4d6c36a8dec4
Simply run
```shell
python test_partition.py
```
You will see result below:
<img width="1620" height="266" alt="Image" src="https://github.com/user-attachments/assets/00aa23fb-49f5-4771-b5b3-ce23c2b96a1b" />

### Versions

Collecting environment information...
PyTorch version: 2.9.0.dev20250906+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: glibc-2.35

Python version: 3.10.12 (main, May 27 2025, 17:12:29) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-151-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  8
On-line CPU(s) list:                     0-7
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) W-2225 CPU @ 4.10GHz
CPU family:                              6
Model:                                   85
Thread(s) per core:                      2
Core(s) per socket:                      4
Socket(s):                               1
Stepping:                                7
CPU max MHz:                             4600.0000
CPU min MHz:                             1200.0000
BogoMIPS:                                8199.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               128 KiB (4 instances)
L1i cache:                               128 KiB (4 instances)
L2 cache:                                4 MiB (4 instances)
L3 cache:                                8.3 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-7
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             KVM: Mitigation: VMX disabled
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsx async abort:           Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] executorch==1.0.0a0+eec95d0
[pip3] mypy==1.17.0
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.2.6
[pip3] onnx==1.19.0
[pip3] onnx-ir==0.1.9
[pip3] onnxscript==0.5.1
[pip3] pytorch_tokenizers==0.1.0
[pip3] torch==2.9.0.dev20250906+cpu
[pip3] torchao==0.14.0+gitb99904b34
[pip3] torchaudio==2.8.0.dev20250906+cpu
[pip3] torchdata==0.11.0
[pip3] torchsr==1.0.4
[pip3] torchtune==0.6.1
[pip3] torchvision==0.24.0.dev20250906+cpu
[conda] Could not collect

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

**Comment by zhxchen17** [comment]
@chenweng-quic Have we filed an issue to downstream Executorch? https://github.com/pytorch/executorch/issues, if I understand this correctly, executorch partitioner seems to be maintained there.

**Comment by cccclai** [comment]
the partitioner is using https://github.com/pytorch/pytorch/blob/5b0b4cda4aa03bee16ee67d9d36012a539df3c50/torch/fx/passes/utils/fuser_utils.py#L95-L102 which is under torch/fx

**Comment by zhxchen17** [comment]
cc @angelayi I saw you commented on the ET repo, do you want to help with this?

**Comment by angelayi** [comment]
@zhxchen17 yeah sure! 
@chenweng-quic sorry for the churn, let's stick to this issue since you created this first (sorry I didn't notice this issue) 😅 Also sorry for the delay in resolving the issues

@chenweng-quic wondering, why does the order of the placeholders changing cause an issue?

**Comment by chenweng-quic** [comment]
> [@chenweng-quic](https://github.com/chenweng-quic) sorry for the churn, let's stick to the ET issue ([pytorch/executorch#14741](https://github.com/pytorch/executorch/issues/14741)) 😅 Also sorry for the delay in resolving the issues

I noticed this issue while investigating some accuracy problems and initially thought it was the root cause. Although I later found out it wasn't, I believe it's still worth fixing this error to avoid any future misunderstandings.

**Comment by SherlockNoMad** [comment]
In fuser_utils.py

```
       sorted_nodes = topo_sort(list(partition))
``` 
should be stable sort, i.e. no change in nodes' relative order.

and `fuse_as_graphmodule` should also preserve original relative order. 

This is important now, as Paritioner is used in the regional inductor workflow. Would appreciate if someone can help fixing this.
